### ISSUE TITLE
Really hard to import topics due to API limitations

### ISSUE THREAD

**Issue by Israphel** [confluentinc/terraform-provider-confluent#105]
Getting these two errors in a row is very common:

`Error: error updating Kafka Topic "lkc-d05x1/usage_v0": reset to topic setting's default value operation (in other words, removing topic settings from 'configs' block) is not supported at the moment. Instead, find its default value at https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic and set its current value to the default value.`

`Error: error updating Kafka Topic "lkc-nvq6yk/usage_v0": "cleanup.policy" topic setting is read-only and cannot be updated. Read https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic for more details.`

You don't allow me use a default config, yet you also don't let me to explicitly enter all the default config because some of them are read only.

I'm ok if they are read only, I'm not even trying to change them, but I just want to import existing topics with existing configs, no changes required. Your API/Provider should be able to detect when I'm changing a config and when I'm not.

Also, topics from dedicated clusters pick these extra config lines:

```
confluent.key.schema.validation: 'false'
confluent.key.subject.name.strategy: 'io.confluent.kafka.serializers.subject.TopicNameStrategy'
confluent.value.schema.validation: 'false'
confluent.value.subject.name.strategy: 'io.confluent.kafka.serializers.subject.TopicNameStrategy'
```

And the same thing happens: They are read only.

**Comment by nick-osborn** [comment]
I feel your pain. :) I had a very interesting time re-importing topics that were created with several items initially left open that we were allowed to create form the start and in turn would either want to update in-place(for the specific items you called out). I ended up keeping a separated list(below) of ones that managed well, didnt manage well, and could not be managed at all in the topic modules that I use. Hopefully it helps. @linouk23, Might not hurt to work in some validation on the "careful with these" settings to either lock them or give some feedback on why the change causes an update-in-place loop(that in my experience always failed).   

  partitions_count = var.num_partitions
    config = {   
    "max.message.bytes"							            =   var.max_message_bytes
    "retention.bytes"							              =   var.retention_bytes
    "retention.ms"								              =   var.retention_ms
    "segment.bytes"								              =   var.segment_bytes    
    "segment.ms"								                =   var.segment_ms
    "max.compaction.lag.ms"						          =   var.max_compaction_lag_ms    
    "min.compaction.lag.ms"						          =   var.min_compaction_lag_ms
    "message.timestamp.difference.max.ms"		    =   var.message_timestamp_difference_max_ms
    "message.timestamp.type"					          =   var.message_timestamp_type
    "delete.retention.ms"						            =   var.delete_retention_ms    
    "min.insync.replicas"						            =   var.min_insync_replicas

    #careful with these
    "cleanup.policy"							              = var.cleanup_policy
    "confluent.key.schema.validation"           = var.confluent_key_schema_validation
    "confluent.key.subject.name.strategy"       = var.confluent_key_subject_name_strategy
    "confluent.value.schema.validation"         = var.confluent_value_schema_validation
    "confluent.value.subject.name.strategy"     = var.confluent_value_subject_name_strategy

    #below are commented - these settings are locked for Confluent Cloud
    # "group.max.session.timeout.ms"				      =   var.group_max_session_timeout_ms
    # "connection.failed.authentication.delay.ms"	=   var.connection_failed_authentication_delay_ms
    # "offsets.retention.minutes"					        =   var.offsets_retention_minutes 
    # "compression.type"							            =   var.compression_type
    # "default.replication.factor"				        =   var.default_replication_factor 
    # "file.delete.delay.ms"						          =   var.file_delete_delay_ms 
    # "flush.messages"							              =   var.flush_messages
    # "flush.ms"									                =   var.flush_ms
    # "index.interval.bytes"						          =   var.index_interval_bytes
    # "message.downconversion.enable"				      =   var.message_downconversion_enable
    # "message.format.version"					          =   var.message_format_version
    # "min.cleanable.dirty.ratio"					        =   var.min_cleanable_dirty_ratio
    # "preallocate"								                =   var.preallocate
    # "segment.index.bytes"						            =   var.segment_index_bytes
    # "segment.jitter.ms"							            =   var.segment_jitter_ms 
    # "unclean.leader.election.enable"			      =   var.unclean_leader_election_enable

**Comment by nick-osborn** [comment]
You also have to be careful with the svc acct that builds these. If I recall correctly, when I made a change to the username of the svc acct it would come up with a destroy + replace that mentioned the endpoint being changed which didnt really make sense to me. I ended up finding out that it was because of some forced truncation of names(to ensure svc acct names didnt go past 64 char limit)

**Comment by Israphel** [comment]
My initial code was like

```terraform
# Topícs
variable "topics" {
  description = "List of topics."
  type = map(object({
    partitions_count = number
    config           = map(string)
  }))
  default = {}
}

# Confluent Kafka Topic
resource "confluent_kafka_topic" "kafka_topic" {
  for_each = var.topics

  topic_name       = each.key
  partitions_count = each.value.partitions_count
  rest_endpoint    = confluent_kafka_cluster.kafka_cluster.rest_endpoint
  config           = each.value.config

  kafka_cluster {
    id = confluent_kafka_cluster.kafka_cluster.id
  }

  credentials {
    key    = confluent_api_key.admin.id
    secret = confluent_api_key.admin.secret
  }
}
```

But I got tired of repeating the same defaults over and over when I call the module, so I tried with the following idea:

```terraform
# Confluent Kafka Topic
resource "confluent_kafka_topic" "kafka_topic" {
  for_each = var.topics

  topic_name       = each.key
  partitions_count = each.value.partitions_count
  rest_endpoint    = confluent_kafka_cluster.kafka_cluster.rest_endpoint
  config = {
    "cleanup.policy"                        = try(each.value.config["cleanup.policy"], "delete")
    "delete.retention.ms"                   = try(each.value.config["delete.retention.ms"], "86400000")
    "max.compaction.lag.ms"                 = try(each.value.config["max.compaction.lag.ms"], "9223372036854775807")
    "max.message.bytes"                     = try(each.value.config["max.message.bytes"], "2097164")
    "message.timestamp.difference.max.ms"   = try(each.value.config["message.timestamp.difference.max.ms"], "9223372036854775807")
    "message.timestamp.type"                = try(each.value.config["message.timestamp.type"], "CreateTime")
    "min.compaction.lag.ms"                 = try(each.value.config["min.compaction.lag.ms"], "0")
    "min.insync.replicas"                   = try(each.value.config["min.insync.replicas"], "2")
    "retention.bytes"                       = try(each.value.config["retention.bytes"], "-1")
    "retention.ms"                          = try(each.value.config["retention.ms"], "604800000")
    "segment.bytes"                         = try(each.value.config["segment.bytes"], "104857600")
    "segment.ms"                            = try(each.value.config["segment.ms"], "604800000")
  }

  kafka_cluster {
    id = confluent_kafka_cluster.kafka_cluster.id
  }

  credentials {
    key    = confluent_api_key.admin.id
    secret = confluent_api_key.admin.secret
  }
}
```

That was fine in most cases, until I hit a topic that wanted "more".

Sadly I had to go back to `config = each.value.config` and then I had to be super repetitive in the YAML I use as the input for my module.

The only good thing is that devs will be able to see all of their configs for every topic, but I would really like to have defaults that work as intended.

**Comment by nick-osborn** [comment]
Interesting, cant say ive used a try before with TF. I think mine is similar in that I set defaults in my module for all the vars that can be used but put the defaults in the variables.tf files. At this rate, especially with this resource I try to not expose much to the dev-side in an effort to prevent some of that and also only give them what the care about at that moment and then give options when necessary. The list I have above I havent hit any issues so far since my last run-in with re-importing topics. You might have an issue with cleanup policy. I encountered weirdness there on one of mine.

**Comment by Israphel** [comment]
yes, cleanup_policy and the four called confluent.something are considered read only. I have only encountered the confluent. configs while importing topics from a dedicated cluster, otherwise they don't exist (in my case).

**Comment by nick-osborn** [comment]
Once managed with TF I would think youre probably good outside of changes on the "weird ones". Is it a pretty common scenario that you will be importing/re-importing topics? It's my opinion(everyone has them right/wrong) that if yore using Terraform in your org that outside of one-off test builds that you would likely want to set guard-rails to only allow creation from TF. Thats what doing in mine(until im overruled). That helps me in a couple of ways. I can force naming conventions(to an extent) with TF - like forcing upper/lower case, regex/replace for any special characters that shouldnt be there, length substr, etc. Just a thought, cause there's not much more that I hate with TF than state migration and imports/removals/re-imports and if thats going to be common might prepare for some pain...

**Comment by Israphel** [comment]
I only need to import once, but the current API limitations like not being able to use "defaults" plus not allowing you to set readonly configs with the very exact value they have now, is preventing me from managing two environments with two regions each with the same list of topics (in YAML format) just because one set (prod us-east-1) happens to be imported and therefore happens to be in a stupid fight against the API.

**Comment by nick-osborn** [comment]
Ohhhh, ok gotcha. I dont currently have multiple clusters so that may be why I havent encountered that. By chance, are you using mirroring for those topics? I havent even checked to see if thats manageable by TF.

**Comment by Israphel** [comment]
No, syncing topics with Confluent dedicated is way too expensive. I will use Amazon MSK for those services that require the same stuff in both regions.

**Comment by linouk23** [comment]
@Israphel thanks for creating this issue! Seems like there're a lot of messages here so I'll try to address them in a bit.

**Comment by linouk23** [comment]
> Also, topics from dedicated clusters pick these extra config lines:
```
confluent.key.schema.validation: 'false'
confluent.key.subject.name.strategy: 'io.confluent.kafka.serializers.subject.TopicNameStrategy'
confluent.value.schema.validation: 'false'
confluent.value.subject.name.strategy: 'io.confluent.kafka.serializers.subject.TopicNameStrategy'
```
> And the same thing happens: They are read only.

We merged a fix for ⬆️ internally already and it'll be available in the next release of TF provider.

**Comment by linouk23** [comment]
> You don't allow me use a default config, yet you also don't let me to explicitly enter all the default config because some of them are read only.
I'm ok if they are read only, I'm not even trying to change them, but I just want to import existing topics with existing configs, no changes required. Your API/Provider should be able to detect when I'm changing a config and when I'm not.

I've just create a new topic on Confluent Cloud UI and managed to import it to TF without any errors:

```
➜  export IMPORT_KAFKA_API_SECRET="..."
➜  export IMPORT_KAFKA_API_KEY="..."                                                   
➜  export IMPORT_KAFKA_REST_ENDPOINT="..." 
➜  terraform import confluent_kafka_topic.test lkc-j930rw/orders
confluent_kafka_topic.test: Importing from ID "lkc-j930rw/orders"...
confluent_kafka_topic.test: Import prepared!
  Prepared confluent_kafka_topic for import
confluent_kafka_topic.test: Refreshing state... [id=lkc-j930rw/orders]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

➜  terraform plan                                               

confluent_kafka_topic.test: Refreshing state... [id=lkc-j930rw/orders]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```

We did add the following note to our [docs](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic#import): 
![image](https://user-images.githubusercontent.com/5459870/189817608-7ef4626d-44e9-42d2-8a00-352a1396df67.png)

Let us know if you think there's a way to improve it.

**Comment by linouk23** [comment]
For the context, there're some API limitations and here's a relevant [piece](https://github.com/confluentinc/terraform-provider-confluent/blob/master/internal/provider/resource_kafka_topic.go#L563) of source code.

In short, API returns either `DYNAMIC_TOPIC_CONFIG` or `DEFAULT_CONFIG` / `STATIC_BROKER_CONFIG` for every topic setting:
```
 {
    "kind": "KafkaTopicConfig",
    "metadata": {
      "self": "https://pkc-xxxx.us-east1.gcp.confluent.cloud/kafka/v3/clusters/lkc-xxxx/topics/topic-a/configs/segment.bytes",
      "resource_name": "crn:///kafka=lkc-xxxx/topic=topic-a/config=segment.bytes"
    },
    "cluster_id": "lkc-xxxx",
    "name": "segment.bytes",
    "value": "536870912",
    "is_read_only": false,
    "is_sensitive": false,
    "source": "DYNAMIC_TOPIC_CONFIG",
    "synonyms": [
      {
        "name": "segment.bytes",
        "value": "536870912",
        "source": "DYNAMIC_TOPIC_CONFIG"
      },
      {
        "name": "log.segment.bytes",
        "value": "104857600",
        "source": "STATIC_BROKER_CONFIG"
      },
      {
        "name": "log.segment.bytes",
        "value": "1073741824",
        "source": "DEFAULT_CONFIG"
      }
    ],
    "topic_name": "topic-a"
  },
```

`DEFAULT_CONFIG` / `STATIC_BROKER_CONFIG` means the topic setting wasn't overridden and `DYNAMIC_TOPIC_CONFIG` means that the topic setting was overridden (if default for a topic setting `foo` is `123` and then someone updates topic setting to `123` again then the resulting value would be `DYNAMIC_TOPIC_CONFIG`.

While ⬆️ works for topics created via CLI / TF, for UI it's even more challenging -- since UI uses the old version of the API to create topics, it returns `DYNAMIC_TOPIC_CONFIG` for all topic settings even if their value are defaults, for example:
```
 {
    "kind": "KafkaTopicConfig",
    "metadata": {
      "self": "https://pkc-xxxx.us-east1.gcp.confluent.cloud/kafka/v3/clusters/lkc-xxxx/topics/topic-a/configs/segment.bytes",
      "resource_name": "crn:///kafka=lkc-xxxx/topic=topic-a/config=segment.bytes"
    },
    "cluster_id": "lkc-xxxx",
    "name": "foo",
    "value": "123",
    "is_read_only": false,
    "is_sensitive": false,
    "source": "DYNAMIC_TOPIC_CONFIG",
    "synonyms": [
      {
        "name": "segment.bytes",
        "value": "123",
        "source": "DYNAMIC_TOPIC_CONFIG"
      },
      {
        "name": "segment.bytes",
        "value": "123",
        "source": "STATIC_BROKER_CONFIG"
      }
    ],
    "topic_name": "topic-a"
  },
```
so we went with the most basic approach where TF marks every `DYNAMIC_TOPIC_CONFIG` as overridden one and prompts a user to provide a value for it (which is the reason why TF asks for every topic setting when importing a Kafka topic created via UI since every topic setting for a topic created via UI has `DYNAMIC_TOPIC_CONFIG`).

That said, we're open to suggestions to make it easier. Let us know if you can see an easier fix.

**Comment by linouk23** [comment]
> By chance, are you using mirroring for those topics? I havent even checked to see if thats manageable by TF.

FWIW adding `confluent_cluster_link` resource is on the roadmap.

**Comment by linouk23** [comment]
> I only need to import once, but the current API limitations like not being able to use "defaults" plus not allowing you to set readonly configs with the very exact value they have now

Let me know if my messages ⬆️ helps with this issue.

**Comment by Israphel** [comment]
The solution should be something like this:

You only specify the few lines you need inside "config" and everything else (the defaults) should be inherited by the provider. And we should be able to add/remove the config block whenever we want.

Also, if we set a readonly config and we don't change the value, then don't return an error. Storing a line with the same value in the state doesn't mean we're changing a readonly value.

There's one specific topic I imported (many times) where the plan tells me it wants to change "cleanup.policy" to "delete". But that topic already has all, and so do the others. I can't put it in my config file because it's also considered a readonly option.

So the results I get when importing are very random, for different topics with the same exact config.

**Comment by lukeknep** [comment]
> No, syncing topics with Confluent dedicated is way too expensive. I will use Amazon MSK for those services that require the same stuff in both regions.

@Israphel In MSK, you would have to use MirrorMaker 2 to sync topic data between clusters. While Confluent Cloud doesn't support Cluster Linking between basic & standard clusters yet, you could still use MirrorMaker 2 or Confluent Replicator to do that between Confluent Cloud clusters. So if you need to sync the same data to multiple regions, you still have a choice between using MSK or basic/standard Confluent Cloud clusters, since they'd use the same tool for replication.

**Comment by Israphel** [comment]
> > No, syncing topics with Confluent dedicated is way too expensive. I will use Amazon MSK for those services that require the same stuff in both regions.
> 
> @Israphel In MSK, you would have to use MirrorMaker 2 to sync topic data between clusters. While Confluent Cloud doesn't support Cluster Linking between basic & standard clusters yet, you could still use MirrorMaker 2 or Confluent Replicator to do that between Confluent Cloud clusters. So if you need to sync the same data to multiple regions, you still have a choice between using MSK or basic/standard Confluent Cloud clusters, since they'd use the same tool for replication.

I know, we already had a chat with Confluent. We made our own tool for syncing MSK, overall it was a better choice than syncing non-dedicated confluent cluster (or paying more dedicated clusters). And we were forced to ditch the VPC peering, on top of that.

**Comment by lukeknep** [comment]
@Israphel Gotcha, thanks for the insight. If you don't mind sharing, why was it a better choice than syncing non-dedicated confluent cloud clusters?

**Comment by Israphel** [comment]
> @Israphel Gotcha, thanks for the insight. If you don't mind sharing, why was it a better choice than syncing non-dedicated confluent cloud clusters?

not sure, a team of devs were assigned to that and they came up with an easier/cheaper solution. This was after several meetings with Confluent.

**Comment by linouk23** [comment]
👋 @Israphel, thanks for sharing the proposed solution!

> You only specify the few lines you need inside "config" and everything else (the defaults) should be inherited by the provider.

I believe that's accurate for topics that are created via TF / CLI. However due to the fact that UI uses currently uses old API a user is forced to list all the topic settings when importing a topic (the issue is described in detail in my previous messages in this comment thread).

> And we should be able to add/remove the config block whenever we want.

Updates for topic settings is supported where possible (see the doc [here](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic)). Topic setting removal doesn't work right now due to bug in the backend (the workaround is to set topic [setting](https://docs.confluent.io/cloud/current/clusters/broker-config.html#custom-topic-settings-for-all-cluster-types) directly to its default value). TF will be updated to support it as soon as the fix is available.

> Also, if we set a readonly config and we don't change the value, then don't return an error. Storing a line with the same value in the state doesn't mean we're changing a readonly value.

That is related to the algorithm we used to implement topic setting updates without topic setting removal support since we didn't want to hardcode any defaults in TF Provider that might change on the backend. As soon as topic setting removal starts working, we might be able to fix it.

> There's one specific topic I imported (many times) where the plan tells me it wants to change "cleanup.policy" to "delete". But that topic already has all, and so do the others. I can't put it in my config file because it's also considered a readonly option.

What happens if you specify `cleanup.policy = delete` before running `terraform import`?

Let me know if it helps.

**Comment by Israphel** [comment]
> What happens if you specify cleanup.policy = delete before running terraform import?

I will try

**Comment by linouk23** [comment]
@nick-osborn speaking about 

> Ohhhh, ok gotcha. I dont currently have multiple clusters so that may be why I havent encountered that. By chance, are you using mirroring for those topics? I havent even checked to see if thats manageable by TF.

We've just released `1.5.0` [version](https://github.com/confluentinc/terraform-provider-confluent/blob/master/CHANGELOG.md#150-september-21-2022) of TF Provider that 
* Adds new confluent_cluster_link [resource](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_cluster_link) and confluent_kafka_mirror_topic [resource](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_mirror_topic) in an [Open Preview lifecycle stage](https://docs.confluent.io/cloud/current/api.html#section/Versioning/API-Lifecycle-Policy).
* Added support for updating [Schema Validation topic settings](https://docs.confluent.io/cloud/current/sr/broker-side-schema-validation.html#sv-configuration-options-on-a-topic) for confluent_kafka_topic [resource](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic).

**Comment by linouk23** [comment]
I was rereading this thread and it seems like `1.5.0` [version](https://github.com/confluentinc/terraform-provider-confluent/blob/master/CHANGELOG.md#150-september-21-2022) of TF Provider that

> Adds support for updating [Schema Validation topic settings](https://docs.confluent.io/cloud/current/sr/broker-side-schema-validation.html#sv-configuration-options-on-a-topic) for confluent_kafka_topic [resource](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_kafka_topic).

should resolve the initial issue that was reported by @Israphel:

![image](https://user-images.githubusercontent.com/5459870/208280050-5f085ee9-04c7-4179-b5f1-f4e2805b32d5.png)

so I'm resolving this issue but feel free to reopen if there're any questions / issues left.
